pd/src/app.rs

```diff
@@ -348,22 +355,31 @@ impl App {
     }
 
     fn end_block(&mut self, end: EndBlock) -> response::EndBlock {
+        let epoch = Epoch::from_blockheight(end.height, self.epoch_duration)
+            .expect("must be able to calculate block epoch");
         self.pending_block
             .as_mut()
             .expect("pending_block must be Some in EndBlock")
             .lock()
             .unwrap()
-            .set_height(end.height);
+            .set_height_and_epoch(end.height, epoch);
 
-        if end.height < 0 {
-            panic!("block height should never be negative");
-        }
+        let is_epoch_boundary = Epoch::is_epoch_boundary(end.height, self.epoch_duration)
+            .expect("unhandleable error determining epoch boundary");
+        // set the flag on the pending block
+        self.pending_block
+            .as_mut()
+            .expect("pending_block must be Some in EndBlock")
+            .lock()
+            .unwrap()
+            .set_epoch_boundary(is_epoch_boundary);
```

Comment:
Here the epoch is set independently from the height (they are set at the same time, but from the point of view of `PendingBlock::set...` they're independent), and the epoch boundary is set separately from either.

I guess this is because the mechanics of computing the epoch and the boundary are done in `Epoch` methods that use the `epoch_duration`, which isn't available inside `PendingBlock`.  But maybe it would be cleaner if `epoch_duration` were passed to `PendingBlock::new`, so that instead of having `set_height_and_epoch` as well as `set_epoch_boundary`, there could just be `set_height`, which would compute the epoch info internally.

This way, there's no possibility of misuse in the API, either the height is set or it isn't, and when it is, the epoch info is guaranteed to be correctly filled in.